clang/lib/3C/CastPlacement.cpp

```diff
@@ -18,107 +18,244 @@
 using namespace clang;
 
 bool CastPlacementVisitor::VisitCallExpr(CallExpr *CE) {
-  auto *FD = dyn_cast_or_null<FunctionDecl>(CE->getCalleeDecl());
-  if (FD && Rewriter::isRewritable(CE->getExprLoc())) {
-    // Get the constraint variable for the function.
-    FVConstraint *FV = Info.getFuncConstraint(FD, Context);
-    // Function has no definition i.e., external function.
+  // Get the constraint variable for the function.
+  Decl *CalleeDecl = CE->getCalleeDecl();
+  FVConstraint *FV = nullptr;
+  FunctionDecl *FD = dyn_cast_or_null<FunctionDecl>(CalleeDecl);
+  if (FD) {
+    FV = Info.getFuncConstraint(FD, Context);
     assert("Function has no definition" && FV != nullptr);
+  } else if (isa_and_nonnull<VarDecl>(CalleeDecl)) {
+    CVarOption Opt = Info.getVariable(cast<VarDecl>(CalleeDecl), Context);
+    if (Opt.hasValue()) {
+      if (isa<PVConstraint>(Opt.getValue()))
+        FV = cast<PVConstraint>(Opt.getValue()).getFV();
+      else
+        FV = dyn_cast<FVConstraint>(&Opt.getValue());
+    }
+  }
+
+  if (FV) {
+    // Now we need to check the type of the arguments and corresponding
+    // parameters to see if any explicit casting is needed.
+    ProgramInfo::CallTypeParamBindingsT TypeVars;
+    if (Info.hasTypeParamBindings(CE, Context))
+      TypeVars = Info.getTypeParamBindings(CE, Context);
 
-    // Did we see this function in another file?
-    auto Fname = FD->getNameAsString();
-    if (!ConstraintResolver::canFunctionBeSkipped(Fname)) {
-      // Now we need to check the type of the arguments and corresponding
-      // parameters to see if any explicit casting is needed.
-      ProgramInfo::CallTypeParamBindingsT TypeVars;
-      if (Info.hasTypeParamBindings(CE, Context))
-        TypeVars = Info.getTypeParamBindings(CE, Context);
-      unsigned PIdx = 0;
-      for (const auto &A : CE->arguments()) {
-        if (PIdx < FD->getNumParams()) {
-          // Avoid adding incorrect casts to generic function arguments by
-          // removing implicit casts when on arguments with a consistently
-          // used generic type.
-          Expr *ArgExpr = A;
+    // Cast on arguments
+    unsigned PIdx = 0;
+    for (const auto &A : CE->arguments()) {
+      if (PIdx < FV->numParams() && (!FD || PIdx < FD->getNumParams())) {
```

Comment:
The check `PIdx < FD->getNumParams()` appears twice. I think the `!FD` is effectively getting ignored.

Reply:
Is it ever possible that `FD->numParams() != FV->numParams()` ?

Reply:
It looks like this happens when the function declaration `FD` is not also a prototype. Then `FD->getNumParams()` returns `0`, so cast insertion on arguments is skipped. This shouldn't be happening. The check here should only be for `FV`.